### ISSUE TITLE
No skipped message in JUnit report for unexpectedly passed tests

### ISSUE THREAD

**Issue by vit0rg** [pytest-dev/pytest#2190]
In 2.9.2 version of pytest for unexpectedly passed tests there was skipped tag with message in JUnit XML report like this:
```xml
<testsuite errors="0" failures="0" name="pytest" skips="1" tests="1" time="6.368">
    <testcase classname="test_example" file="test_example.py" line="127" name="test_example" time="0.22827982902526855">
        <skipped message="xfail-marked test passes unexpectedly">None</skipped>
        <system-out>
        </system-out>
    </testcase>
</testsuite>
```

But since version 3.0.0 skipped tag has gone:
```xml
<testsuite errors="0" failures="0" name="pytest" skips="1" tests="1" time="6.368">
    <testcase classname="test_example" file="test_example.py" line="127" name="test_example" time="0.22827982902526855">
        <system-out>
        </system-out>
    </testcase>
</testsuite>
```

So for now there is no way to distinguish passed and unexpectedly passed tests.

**Comment by nicoddemus** [comment]
Thanks for taking the time to write this issue.

This ended up changed for `3.0.0` as an effect of #1795 in response to #1546.

As the test was marked with `xfail`, and not with `skip` or `skipif`, I don't think it is correct to mark it as `skipped` when it xpasses. Unfortunately JUnit XML does not have the notion of "unexpectedly passed".

**Comment by vit0rg** [comment]
Hello nicoddemus, the problem isn't about skipped or succeed, but about that before 3.0.0 there was at least some way to check if test is passed unexpectedly and now there is no such way, so adapted (support XPASS) report systems lies and can't report about XPASS tests which is pretty much problem.

**Comment by RonnyPfannschmidt** [comment]
@vit0rg is setting xpass to strict via pytest.ini an viable option for you?

back when i was made aware of the major mess up with non-strict xfail it was already released - its hard to undo such blunders

**Comment by vit0rg** [comment]
For those who affected, i've done small monkey patch even with little fix for old bug - no reason in JUnit XML report for XPASS in my reporting package:
```python
# Nasty monkey patch for JUnit reporting that doesn't unexpectedly passed tests. See
# https://github.com/pytest-dev/pytest/issues/2190 for details.
def append_pass(self, report):
    if hasattr(report, "wasxfail"):
        self._add_simple(Junit.skipped, "xfail-marked test passes unexpectedly", report.wasxfail)
    self.add_stats('passed')
    self._write_captured_output(report)


_NodeReporter.append_pass = append_pass
```

**Comment by vit0rg** [comment]
> @vit0rg is setting xpass to strict via pytest.ini an viable option for you?
Was my first hope, but doesn't helps because if pytest JUnit reporting implementation - it adds specific message only in case when test is skipped.

**Comment by nicoddemus** [comment]
> the problem isn't about skipped or succeed, but about that before 3.0.0 there was at least some way to check if test is passed unexpectedly and now there is no such way, so adapted (support XPASS) report systems lies and can't report about XPASS tests which is pretty much problem.

@vit0rg I appreciate this change broke your reporting setup and that was a risk we were aware when made that change, which is one of the reasons why it was rolled into 3.0. 

Unfortunately the old behavior likely won't be included back, because it was not correct in the first place, after all "skipped" means the test was not executed at all (usually because of some missing dependency, platform mismatch etc), while an "xpassed" test executed but passed unexpectedly. Also the previous behavior also had the drawback of not being able to differentiate from `xpass` and `skipped` either.

Is the report system you mentioned maintained by you? Because perhaps you can use the [record-xml-property](http://doc.pytest.org/en/latest/usage.html?highlight=record_xml_property#record-xml-property) fixture to attach a custom property to identify the xpassed tests.

**Comment by vit0rg** [comment]
> Is the report system you mentioned maintained by you? Because perhaps you can use the record-xml-property fixture to attach a custom property to identify the xpassed tests.

Yep, reporting system maintained by me, so for me it's easy to support any meta tag or attribute/property, actually current implementation for JUnit XML isn't good because it checks message in skipped tag, so adding meta-property is more correct way, but adding record_xml_property fixture to all tests isn't good idea, so i will look forward to make it globally in some conftest.py event and yet will use monkey patch for _pytest.junitxml._NodeReporter.

**Comment by RonnyPfannschmidt** [comment]
it may be easier to just copy the junitxml plugin and alter it to fit your needs

personally i'd also love to move junitxml to a plugin at some point, because having it in the core is a pain and lock-in

**Comment by nicoddemus** [comment]
> but adding record_xml_property fixture to all tests isn't good idea

You don't have to, you can add an `autouse` fixture to your top `conftest.py` and it will be applied to all your tests automatically:

```python
@pytest.fixture(autouse=True)
def apply_skipped_tag(record_xml_property, request):
    yield
    if hasattr(report, "wasxfail"):
         record_xml_property('was-skipped', True)
```

**Comment by RonnyPfannschmidt** [comment]
for more robust code it makes sense to use `getattr(request.item, "wasxfail", False)` to avoid the exception eating race condition of hasattr

**Comment by vit0rg** [comment]
> but adding record_xml_property fixture to all tests isn't good idea
> You don't have to, you can add an autouse fixture to your top conftest.py and it will be applied to all your tests automatically:

Thanks, but yet i can't find way to access report from within fixture. It should be somewhere in request, but i can't find it.

**Comment by nicoddemus** [comment]
Oh right, xfail information is kept in the `report`, not in `request`... :/

**Comment by webknjaz** [comment]
ref https://github.com/pytest-dev/pytest/issues/7327
